src/desktop/space.rs

```diff
@@ -0,0 +1,419 @@
+        panic!("Out of space ids");
+    }
+
+    let mut id = SPACE_ID.fetch_add(1, Ordering::SeqCst);
+    while ids.iter().any(|k| *k == id) {
+        id = SPACE_ID.fetch_add(1, Ordering::SeqCst);
+    }
+
+    ids.insert(id);
+    id
+}
+
+#[derive(Default)]
+struct WindowState {
+    location: Point<i32, Logical>,
+    drawn: bool,
+}
+
+type WindowUserdata = RefCell<HashMap<usize, WindowState>>;
+fn window_state(space: usize, w: &Window) -> RefMut<'_, WindowState> {
+    let userdata = w.user_data();
+    userdata.insert_if_missing(WindowUserdata::default);
+    RefMut::map(userdata.get::<WindowUserdata>().unwrap().borrow_mut(), |m| {
+        m.entry(space).or_default()
+    })
+}
+
+#[derive(Clone, Default)]
+struct OutputState {
+    location: Point<i32, Logical>,
+    render_scale: f64,
+    // damage and last_state in space coordinate space
+    old_damage: VecDeque<Vec<Rectangle<i32, Logical>>>,
+    last_state: IndexMap<usize, Rectangle<i32, Logical>>,
+}
+
+type OutputUserdata = RefCell<HashMap<usize, OutputState>>;
+fn output_state(space: usize, o: &Output) -> RefMut<'_, OutputState> {
+    let userdata = o.user_data();
+    userdata.insert_if_missing(OutputUserdata::default);
+    RefMut::map(userdata.get::<OutputUserdata>().unwrap().borrow_mut(), |m| {
+        m.entry(space).or_default()
+    })
+}
+
+// TODO: Maybe replace UnmanagedResource if nothing else comes up?
+#[derive(Debug, thiserror::Error)]
+pub enum SpaceError {
+    #[error("Window is not mapped to this space")]
+    UnknownWindow,
+}
+
+#[derive(Debug)]
+pub struct Space {
+    id: usize,
+    // in z-order, back to front
+    windows: IndexSet<Window>,
+    outputs: Vec<Output>,
+    // TODO:
+    //layers: Vec<Layer>,
+    logger: ::slog::Logger,
+}
+
+impl Drop for Space {
+    fn drop(&mut self) {
+        SPACE_IDS.lock().unwrap().remove(&self.id);
+    }
+}
+
+impl Space {
+    pub fn new<L>(log: L) -> Space
+    where
+        L: Into<slog::Logger>,
+    {
+        Space {
+            id: next_space_id(),
+            windows: IndexSet::new(),
+            outputs: Vec::new(),
+            logger: log.into(),
+        }
+    }
+
+    /// Map window and moves it to top of the stack
+    ///
+    /// This can safely be called on an already mapped window
+    pub fn map_window(&mut self, window: &Window, location: Point<i32, Logical>) -> Result<(), SpaceError> {
+        window_state(self.id, window).location = location;
+        self.windows.shift_remove(window);
+        self.windows.insert(window.clone());
+        Ok(())
+    }
+
+    pub fn raise_window(&mut self, window: &Window) {
+        let loc = window_state(self.id, window).location;
+        let _ = self.map_window(window, loc);
+
+        // TODO: should this be handled by us?
+        window.set_activated(true);
+        for w in self.windows.iter() {
+            if w != window {
+                w.set_activated(false);
+            }
+        }
```

Comment:
I'm slightly leaning towards "yes"

Reply:
I was wondering, if we should add a `activate`-boolean to `map_window` (and possible `raise_window`). Downstream might not always want to shift focus, just because a new window is spawned. Even though this does not actually change the focus, it looks wrong.